tests/test_files_cloud.py

```diff
@@ -0,0 +1,99 @@
+"""
+This module contains integration tests which verify the behavior of the storage
+module against supported cloud storage providers. The tests for the local
+(directory-based) storage backend are re-used and run for each supported cloud
+storage provider. To add tests for a new provider, create a new subclass of
+CloudTestFile.
+
+To run tests for Google Cloud Platform, set the following environment variables:
+- GCP_STORAGE_KEY
+- GCP_STORAGE_SECRET
+
+"""
+# TODO(@colinschoen) Configure CI to run the full integration test suite only on protected branches like master.
+
+import os
+import random
+import unittest
+
+import requests
+
+from server.extensions import storage
+from tests.test_files import TestFile
+
+
+class EnvironmentMutationMixin(object):
+    __env_backup = {}
+
+    @classmethod
+    def set_environment_variable(cls, key, value, keep_existing=False):
+        old_value = os.environ.get(key)
+        cls.__env_backup.setdefault(key, old_value)
+        if not old_value or not keep_existing:
+            os.environ[key] = value
+
+    @classmethod
+    def restore_environment_variable(cls, key):
+        original_value = cls.__env_backup.get(key)
+
+        if original_value is None:
+            del os.environ[key]
+        else:
+            os.environ[key] = original_value
+
+
+class CloudTestFile(TestFile, EnvironmentMutationMixin):
+    storage_provider = ""
+    key_env_name = ""
+    secret_env_name = ""
+
+    _storage_container = "okpycloudfilestest{}".format(random.randint(0, 100000))
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        storage_key = os.environ.get(cls.key_env_name)
+        storage_secret = os.environ.get(cls.secret_env_name)
+
+        if not storage_key or not storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for {} not configured".format(cls.storage_provider))
+
+        cls.set_environment_variable("STORAGE_PROVIDER", cls.storage_provider)
+        cls.set_environment_variable("STORAGE_KEY", storage_key)
+        cls.set_environment_variable("STORAGE_SECRET", storage_secret)
+        cls.set_environment_variable("STORAGE_CONTAINER", cls._storage_container, keep_existing=True)
+
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+
+        cls.restore_environment_variable("STORAGE_PROVIDER")
+        cls.restore_environment_variable("STORAGE_KEY")
+        cls.restore_environment_variable("STORAGE_SECRET")
+        cls.restore_environment_variable("STORAGE_CONTAINER")
+
+        container = storage.driver.get_container(cls._storage_container)
```

Comment:
This will throw a `ContainerDoesNotExistError` when `STORAGE_CONTAINER` is set by the user.

Reply:
Thanks for the continued thorough review, @colinschoen.

At this point, I think it's simpler if we just require the storage container to be already created by the time that the integration tests run, so I've implemented that in f1b2e38. We can re-use the same container for all test runs so it's only a one-time setup cost.

Requiring a pre-existing container simplifies handling of the environment variables and also means that we don't require a special service account with create/delete bucket permissions to run the tests.

Note that I've introduced a new environment variable for the storage account name used in the integration tests so that a container name for the local tests can be set via `CONTAINER_NAME` and a different container name for the the GCP cloud storage integration test can be set via `GCP_STORAGE_CONTAINER` (to prevent potential clashes between names that are valid local containers but not in the cloud provider).

Here is how I ran the tests:

1. Create a new Storage project in GCP.
2. Set the project as the default project for interoperable access.
3. Copy the access key to `GCP_STORAGE_KEY` and the secret to `GCP_STORAGE_SECRET`.
4. Create a new bucket and copy its name to `GCP_STORAGE_CONTAINER`.
5. Execute `python manage.py test` (passes).
6. Unset all the `GCP_` environment variables.
7. Run `python manage.py test` (passes).

Thoughts?

Reply:
Awesome, I like this a lot. I will give it a go.

Reply:
Haven't had a chance to debug what is going on, but I am failing a few `GoogleCloudTestFile` tests as well as `TestFile` tests when setting the following environment variables: `GCP_STORAGE_KEY`, `GCP_STORAGE_SECRET`, `GCP_STORAGE_CONTAINER`. 

Interestingly, I don't fail those `TestFile` tests when the said environment variables are unset. 

I will be taking a look to figure out why. 
[test_output.txt](https://github.com/Cal-CS-61A-Staff/ok/files/1940632/test_output.txt)